core/src/main/java/org/jruby/truffle/nodes/rubinius/TimePrimitiveNodes.java

```diff
@@ -0,0 +1,316 @@
+
+        public TimeDecomposePrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            toDateTimeNode = new RubyTimeToDateTimeNode(context, sourceSection);
+        }
+
+        public TimeDecomposePrimitiveNode(TimeDecomposePrimitiveNode prev) {
+            super(prev);
+            toDateTimeNode = prev.toDateTimeNode;
+        }
+
+        @Specialization
+        public RubyArray timeDecompose(VirtualFrame frame, RubyTime time) {
+            final DateTime dateTime = toDateTimeNode.toDateTime(frame, time);
+            final Object[] decomposed = decompose(dateTime);
+            return new RubyArray(getContext().getCoreLibrary().getArrayClass(), decomposed, decomposed.length);
+        }
+
+        @CompilerDirectives.TruffleBoundary
+        public Object[] decompose(DateTime dateTime) {
+            final int sec = dateTime.getSecondOfMinute();
+            final int min = dateTime.getMinuteOfDay();
+            final int hour = dateTime.getHourOfDay();
+            final int day = dateTime.getDayOfMonth();
+            final int month = dateTime.getMonthOfYear();
+            final int year = dateTime.getYear();
+            final int wday = dateTime.getDayOfWeek();
+            final int yday = dateTime.getDayOfYear();
+            final Object isdst = getContext().getCoreLibrary().getNilObject();
+            final Object zone = getContext().getCoreLibrary().getNilObject();
+            return new Object[]{sec, min, hour, day, month, year, wday, yday, isdst, zone};
+        }
+
+    }
+
+    @RubiniusPrimitive(name = "time_strftime")
+    public static abstract class TimeStrftimePrimitiveNode extends RubiniusPrimitiveNode {
+
+        @Child protected RubyTimeToDateTimeNode toDateTimeNode;
+
+        public TimeStrftimePrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            toDateTimeNode = new RubyTimeToDateTimeNode(context, sourceSection);
+        }
+
+        public TimeStrftimePrimitiveNode(TimeStrftimePrimitiveNode prev) {
+            super(prev);
+            toDateTimeNode = prev.toDateTimeNode;
+        }
+
+        @Specialization
+        public RubyString timeStrftime(VirtualFrame frame, RubyTime time, RubyString format) {
+            return getContext().makeString(format(toDateTimeNode.toDateTime(frame, time), time.getNanoseconds(), format.getBytes()));
+        }
+
+
+        @CompilerDirectives.TruffleBoundary
+        public ByteList format(DateTime time, long nanoseconds, ByteList format) {
+            // TODO: converts everything to JRuby objects and back - should find a more direct way using ByteList
+            final RubyDateFormatter rdf = getContext().getRuntime().getCurrentContext().getRubyDateFormatter();
+            return rdf.compileAndFormat(org.jruby.RubyString.newString(getContext().getRuntime(), format), false, time, nanoseconds, null).getByteList();
```

Comment:
I did a lot of work on this, I'll need to check why I kept using a RubyString as input.